### PR TITLE
Route leak feature

### PR DESCRIPTION
**Pull request by kentwu1**
1. defined 'apic:allow_route_leak' as the network extension attribute.
2. when a route_leak network is connected to a router under per router vrf mode,
the corresponding leak_epg, leak_bd and the contracts will be created accordingly
also on APIC.
3. added a new route_leak_list section in the rdconfig file. This along with item 2
will tell opflex agent on the computes that from which source to which destination
subnets should this route leak rules apply. Basically the opflex agent should then
do another ep_declare also on the leak_epg when an ep file is created on the route
leak network.
4. also calculated the right internal-subnets for the rdconfig file when a route_leak
network is being involved.

### CODE REVIEW

**Comment by kentwu1** [comment]
pushed the new revision to use the l3out instead of rdconfig approach.

File: apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py [hunk 1]
```diff
@@ -598,8 +600,21 @@ def get_vrf_details(self, context, **kwargs):
...
                 nets = set([p['network_id'] for p in intf_ports])
             else:
-                nets = core_plugin.get_networks(ctx, {'tenant_id': [vrf_id]})
-                nets = [n['id'] for n in nets]
+                networks = core_plugin.get_networks(ctx,
+                                                    {'tenant_id': [vrf_id]})
+                nets = [n['id'] for n in networks]
+                obj = {}
+                obj['tenant_id'] = vrf_id
+                # for route leak case
```
Comment (amittbose) [inline apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py]: So is this code block trying to excluding non-leaked routed subnets from the 'shared' VRF? If so, please add a comment to that effect and why that is needed

File: apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py [hunk 2]
```diff
@@ -908,6 +924,32 @@ def _add_ip_mapping_details(self, context, port, host, owned_addr,
...
+                         [n_constants.DEVICE_OWNER_ROUTER_INTF],
+                         'network_id': [port_network['id']]})
+            port_sbunet = self._get_port_subnets(port).pop()
+            router_ids = [x['device_id'] for x in intf_ports
+                          if x['fixed_ips'][0]['subnet_id'] == port_sbunet]
+            fixed_ips = [x['ip_address'] for x in port['fixed_ips']]
+            epg_tenant = self.apic_manager.apic.fvTenant.name(
+                self._get_tenant(port_network))
+            for router_id in router_ids:
+                fip = {'id': str(uuid.uuid4()),
```
Comment (amittbose) [inline apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py]: can we avoid a generating a new uuid during each invocation of this call, perhaps by using port id + router id? Changing uuids can cause unnecessary churn on the agent and leaf switch?

File: apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py [hunk 3]
```diff
@@ -1046,24 +1094,155 @@ def _perform_interface_port_operations(self, context, port, network,
...
+                    intf_net_ports = context._plugin.get_ports(
                         context._plugin_context,
                         filters={'device_owner':
                                  [n_constants.DEVICE_OWNER_ROUTER_INTF],
-                                 'device_id': [router['id']],
+                                 'device_id': [router_id],
                                  'network_id': [network['id']]})
-                    if not [p for p in intf_ports if p['id'] != port['id']]:
-                        vrf_default = self._get_network_vrf(context, network)
+                    if not [p for p in intf_net_ports
```
Comment (amittbose) [inline apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py:1200]: this function has become too long, please move handling of the route-leak external-epg to separate functions that can be called from here.

**Comment by kentwu1** [comment]
just pushed a new revision to address those review comments

**Review by amittbose — APPROVED** [review]

**Pull request merged**